crates/ra_proc_macro/src/lib.rs

```diff
@@ -0,0 +1,78 @@
+//! Client-side Proc-Macro crate
+//!
+//! We separate proc-macro expanding logic to an extern program to allow
+//! different implementations (e.g. wasm or dylib loading). And this crate
+//! is used for provide basic infra-structure for commnicate between two
+//! process: Client (RA itself), Server (the external program)
+
+use ra_mbe::ExpandError;
+use ra_tt::Subtree;
+use std::{
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+trait ProcMacroExpander: std::fmt::Debug + Send + Sync + std::panic::RefUnwindSafe {
+    fn custom_derive(&self, subtree: &Subtree, derive_name: &str) -> Result<Subtree, ExpandError>;
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct ProcMacroProcessExpander {
+    process_path: PathBuf,
+}
+
+impl ProcMacroExpander for ProcMacroProcessExpander {
+    fn custom_derive(
+        &self,
+        _subtree: &Subtree,
+        _derive_name: &str,
+    ) -> Result<Subtree, ExpandError> {
+        // FIXME: do nothing for now
+        Ok(Subtree::default())
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ProcMacro {
+    expander: Arc<Box<dyn ProcMacroExpander>>,
```

Comment:
What is the rationale for `Arc<Box<T>>`? Maybe just `Arc<T>` ?

Reply:
Nice catch. I just forget to change it back. thanks!